samples/boards/nrf/s2ram/src/main.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2022 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <pm/pm.h>
+#include <pm/policy.h>
+
+#include <arch/common/pm_s2ram.h>
+
+#include <hal/nrf_vmc.h>
+#include <hal/nrf_power.h>
+#include <hal/nrf_regulators.h>
+#include <hal/nrf_gpio.h>
+
+#define RAM_N_BLOCK	(8)
+#define MASK_ALL_SECT	(VMC_RAM_POWER_S0RETENTION_Msk  | VMC_RAM_POWER_S1RETENTION_Msk  | \
+			 VMC_RAM_POWER_S2RETENTION_Msk  | VMC_RAM_POWER_S3RETENTION_Msk  | \
+			 VMC_RAM_POWER_S4RETENTION_Msk  | VMC_RAM_POWER_S5RETENTION_Msk  | \
+			 VMC_RAM_POWER_S6RETENTION_Msk  | VMC_RAM_POWER_S7RETENTION_Msk  | \
+			 VMC_RAM_POWER_S8RETENTION_Msk  | VMC_RAM_POWER_S9RETENTION_Msk  | \
+			 VMC_RAM_POWER_S10RETENTION_Msk | VMC_RAM_POWER_S11RETENTION_Msk | \
+			 VMC_RAM_POWER_S12RETENTION_Msk | VMC_RAM_POWER_S13RETENTION_Msk | \
+			 VMC_RAM_POWER_S14RETENTION_Msk | VMC_RAM_POWER_S15RETENTION_Msk)
+
+static void ram_retention_config(bool enable)
+{
+	uint32_t mask_sect = MASK_ALL_SECT;
+
+	for (size_t n = 0; n < RAM_N_BLOCK; n++) {
+		if (enable) {
+			nrf_vmc_ram_block_retention_set(NRF_VMC, n, mask_sect);
+		} else {
+			nrf_vmc_ram_block_retention_clear(NRF_VMC, n, mask_sect);
+		}
+	}
+}
+
+static void system_off(void)
+{
+	nrf_regulators_system_off(NRF_REGULATORS);
+}
+
+static void do_suspend(void)
+{
+	irq_lock();
+
+	/*
+	 * Save the CPU context (including the return address),set the SRAM
+	 * marker and power off the system.
+	 */
+	arch_pm_s2ram_suspend(system_off);
```

Comment:
if applications trigger s2ram, the `PM_STATE_SUSPEND_TO_RAM` should probably be skipped by system PM code (it will be returned by the policy manager if defined in DT and forwarded to `pm_state_set()`, an SoC specific call.